int32 index
bool rotationDirection